### ISSUE TITLE
flipper_id required for "thing" in group?

### ISSUE THREAD

**Issue by LaurelOlson** [jnunemaker/flipper#402]
From the [docs](https://github.com/jnunemaker/flipper/blob/master/docs/Gates.md#5-group):

> There is no requirement that the thing yielded to the block be a user model or whatever. It can be anything you want, therefore it is a good idea to check that the thing passed into the group block actually responds to what you are trying to do in the register proc.

However, given a group that is meant to take in a string as the "thing yielded to the block", the "must respond to flipper_id" error is thrown when running `Flipper[:feature].enabled?("thing")`. Looking through the code, the `enabled?` method [tries to instantiate an actor](https://github.com/jnunemaker/flipper/blob/7623cd13e836fc55de8f31607d573161b1c865e4/lib/flipper/feature.rb#L106), which then [throws the error](https://github.com/jnunemaker/flipper/blob/a3a318dd65fcdfbc60a43f3677c6e0d85bfc40e5/lib/flipper/types/actor.rb#L14) due to the missing `flipper_id` method.

Is there a way to check if a "thing" is enabled specifically as a group member without requiring that it responds to `flipper_id`?

**Comment by AlexWheeler** [comment]
Since `Flipper[:feature].enabled?(obj)` checks all the gate types, with Groups being one of them, to determine if a feature is enabled for the obj, it needs to respond to flipper_id.

Groups are great for enabling lots of actors or when you want to enable different types of actors (account, user, etc.)

It could be helpful if you pasted your code with your actor to see what you're trying to accomplish, as you might be better off using an actor abstraction like Struct.new(:flipper_id).new("thing") instead of registering a group (although beware flipper_id must be unique).

**Comment by LaurelOlson** [comment]
OK thanks, yea that's what I figured. I tried converting the actors to an OpenStruct with a flipper_id and that seems to be working. The use case is to enable/disable features per mobile app version which we've made work by doing something like this:

```
Flipper.register("ios-2.4.1-744+") do |actor| 
     actor.os == "ios" && actor.version >= "2.4.1" && actor.build_number >= 744
end
```

where an actor would be `OpenStruct.new({ os: "ios", version: "2.4.1", build_number: "744", flipper_id: "ios-2.4.1-744"})`

**Comment by jnunemaker** [comment]
[Flipper::Actor](https://github.com/jnunemaker/flipper/blob/1bfb71c53ddd2097364a8c1d3ccfd61752399497/lib/flipper/actor.rb) is great for this.

Also, from the example above, you might want a MobileRequest class. It could accept a hash of parameters and define a flipper_id method that merges those together as you showed. It could also take a request instead and be more like a decorator. We did both of these at GitHub and it worked well. 

The main reason I didn't accept a string directly is kind of a dynamic vs static typing thing. I think for libraries like flipper it is better to be specific as long as it doesn't add much work. For the most part the only work requiring a given interface (like flipper_id) adds is needing a few more classes and instances of those classes which usually leads to greater clarity in the long run.

Closing this issue to keep things tidy, but feel free to keep conversing and we'll keep responding.

Also, just curious, but are you using this to control features for mobile? Very cool if so. I'm all ears if you want to explain more about what you are doing. If not, no worries.
